### PR TITLE
Fully switch unit tests over to Smarty5

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
We are still running Gencode using Smarty4 - which conflicts with Smarty5 in the unit test context - this fixes to fully use Smarty5

Before
----------------------------------------
Potential mix of Smarty4 & Smarty5

After
----------------------------------------
Smarty5

Technical Details
----------------------------------------
Alternative to https://github.com/civicrm/civicrm-core/pull/30640

Comments
----------------------------------------

### CODE REVIEW

**Comment by seamuslee001** [comment]
@eileenmcnaughton 

```
Deprecated: Using Smarty::getPluginsDir() is deprecated and will be removed in a future release. It will always return an empty array. in /home/homer/buildkit/build/build-4/web/sites/all/modules/civicrm/packages/smarty5/vendor/smarty/smarty/src/Smarty.php on line 812
```

**Comment by eileenmcnaughton** [comment]
ok - yeah - 2 places where the smarty version is hard-coded - lessee

File: CRM/Core/CodeGen/Util/Smarty.php [hunk 1]
```diff
@@ -48,28 +48,16 @@ public function getCompileDir() {
   public function createSmarty(): Smarty {
     $base = dirname(__DIR__, 4);
     $pkgs = file_exists(dirname($base) . "/civicrm-packages") ? dirname($base) . "/civicrm-packages" : "$base/packages";
-    require_once $pkgs . '/smarty4/vendor/autoload.php';
+    require_once $pkgs . '/smarty5/Smarty.php';
     $smarty = new Smarty();
     $smarty->setTemplateDir("$base/xml/templates");
     $pluginsDirectory = $smarty->getPluginsDir();
```
Comment (seamuslee001) [inline CRM/Core/CodeGen/Util/Smarty.php:54]: @eileenmcnaughton doesn't this need to go as well?

Comment (eileenmcnaughton) [reply CRM/Core/CodeGen/Util/Smarty.php]: @seamuslee001 but how else would we set the variable that we don't use?

**Comment by seamuslee001** [comment]
```
Fatal error: Cannot redeclare smarty_ucfirst_ascii() (previously declared in /home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/packages/smarty4/vendor/smarty/smarty/libs/functions.php:23) in /home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/packages/smarty5/vendor/smarty/smarty/src/functions.php on line 21
```

File: templates/CRM/common/civicrm.settings.php.template [hunk 2]
```diff
@@ -261,7 +261,7 @@ if (!defined('CIVICRM_UF_BASEURL')) {
  */
 if (!defined('CIVICRM_SMARTY_AUTOLOAD_PATH') && getenv('CIVICRM_SMARTY_AUTOLOAD_PATH') === FALSE) {
   // enable by default for dev & demo sites for now - will soon enable by default for all new installs.
-  if (strpos(CIVICRM_UF_BASEURL, 'localhost') !== FALSE || strpos(CIVICRM_UF_BASEURL, 'demo.civicrm.org') !== FALSE) {
+  if (CIVICRM_UF === 'UnitTests' || strpos(CIVICRM_UF_BASEURL, 'localhost') !== FALSE || strpos(CIVICRM_UF_BASEURL, 'demo.civicrm.org') !== FALSE) {
```
Comment (seamuslee001) [inline templates/CRM/common/civicrm.settings.php.template:264]: I think we should add in here `|| strpos(CIVICRM_UF_BASEURL, 'nip.io') !== FALSE`

cc @demeritcowboy

Comment (eileenmcnaughton) [reply templates/CRM/common/civicrm.settings.php.template]: @seamuslee001 yeah - I guess that would work - I grabbed the string `http://build` which I think would also do it & *might* capture more buildkit installs? (we are pretty close to switching 5 to being the default so a bit greedy is OK I think)

**Comment by highfalutin** [comment]
Seems to fix my original problem.

**Comment by JoeMurray** [comment]
@eileenmcnaughton should get a callout in the dev digest for pulling the codebase screaming and emitting errors all over the place from insecure obsolete Smarty2 to current secure and supported Smarty5 ! Woot! With a little help from her friends of course.

**Comment by twomice** [comment]
Not sure why, but this commit  appears to break `civibuild reinstall` on my drupal-clean.  If I checkout the commit just previous, the rebuild goes fine, but this commit and subsequent ones lead to this error outpout:

```
...
civicrm_domain.version := 5.77.alpha1

Generating civicrm.config.php
Generating sql file
Generating sql drop tables file
Generating data files for en_US
PHP Warning:  require_once(/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/packages/smarty5/Smarty.php): failed to open stream: No such file or directory in /opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Util/Smarty.php on line 51
PHP Fatal error:  require_once(): Failed opening required '/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/packages/smarty5/Smarty.php' (include_path='.:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/packages:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/tecnickcom:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/auth_sasl:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/console_getopt:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/db:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/log:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/mail:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/mail_mime:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/net_smtp:/opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/vendor/pear/ne in /opt/buildkit/build/drupal-clean/web/sites/all/modules/civicrm/CRM/Core/CodeGen/Util/Smarty.php on line 51

```

Perhaps it's something about my local setup. Anyone else seeing this?

**Comment by eileenmcnaughton** [comment]
and is it correct that the file does not exist?

**Comment by seamuslee001** [comment]
@twomice can you check you have the latest master in your packages folder as well just in case

**Comment by twomice** [comment]
@eileenmcnaughton Yes, the file was indeed not present, but @seamuslee001 hit on the answer WRT to the civicrm-packages repo. A simple `git pull` on that repo solved the problem. Silly me. Thanks!

**Pull request merged**